middleman_protocol/middleman_protocol/helpers.py

```diff
@@ -0,0 +1,81 @@
+from middleman_protocol.constants import ESCAPE_SEQUENCES
+from middleman_protocol.constants import FRAME_PAYLOAD_LENGTH_STARTING_BYTE
+from middleman_protocol.constants import FRAME_PAYLOAD_STARTING_BYTE
+from middleman_protocol.constants import FRAME_SEPARATOR
+from middleman_protocol.constants import FRAME_SEPARATOR_LENGTH
+from middleman_protocol.constants import RECEIVE_BYTES_PER_LOOP
+from middleman_protocol.message import AbstractMiddlemanMessage
+
+
+def handle_frame_receive(connection: socket.socket, separator_escaping_expected: bool=False) -> Iterable[bytes]:
+    """
+    Helper generator for receiving serialized Middleman protocol messages.
+
+    Yields received bytes.
+    """
+    assert isinstance(connection, socket.socket)
+    assert isinstance(separator_escaping_expected, bool)
+
+    received_data = b''
+    header_plus_payload_length = None
+    # Used to mark how many bytes was received since last decoding, to avoid decoding twice same part of data.
+    decoding_buffer_length = 0
+
+    while True:
+        received_data += connection.recv(RECEIVE_BYTES_PER_LOOP)
+
+        # If separator is met, reset received data and restart.
+        if (
+            len(received_data) >= FRAME_SEPARATOR_LENGTH and
+            received_data[-FRAME_SEPARATOR_LENGTH:] == FRAME_SEPARATOR
+        ):
+            received_data = b''
+            header_plus_payload_length = None
+            decoding_buffer_length = 0
+            continue
+
+        # If encoding is expected to be used by data sender,
+        # and there is encoded sequence in last received data,
+        # decode it and set encoding buffer.
+        if (
+            separator_escaping_expected and (
+                ESCAPE_SEQUENCES[FRAME_SEPARATOR] in received_data[decoding_buffer_length:] or
+                ESCAPE_SEQUENCES[ESCAPE_CHARACTER] in received_data[decoding_buffer_length:]
+            )
+        ):
+            received_data = (
+                received_data[:decoding_buffer_length] +
+                AbstractMiddlemanMessage.escape_decode_raw_message(received_data[decoding_buffer_length:])
+            )
+            decoding_buffer_length = len(received_data)
+
+        # When defined amount of bytes is received, read payload length.
+        # There is a range defined, because it might happen that payload length matches part of one of escape sequences,
+        # so loop should read it completely first, then decode, and then finally read payload length.
+        if FRAME_PAYLOAD_STARTING_BYTE <= len(received_data) <= (FRAME_PAYLOAD_STARTING_BYTE + len(ESCAPE_CHARACTER)):
+            # Last received data must not match start of ESCAPE_CHARACTER,
+            # otherwise ESCAPE_CHARACTER should be read completely before reading payload length.
+            if any([
+                ESCAPE_CHARACTER.startswith(received_data[-i:])
+                for i in range(len(ESCAPE_CHARACTER))
+            ]):
```

Comment:
It should be `range(1, max(len(ESCAPE_CHARACTER), len(received_data)))`

- Without `1` it unnecessarily does `ESCAPE_CHARACTER.startswith(received_data[0:])` which is the same as ESCAPE_CHARACTER.startswith(received_data).
- The `max()` makes it stop comparing if `received_data` is shorter than the `ESCAPE_CHARACTER`.

Reply:
It will be changed even more when we switch to single character.